Add UTF-8 validation fast paths in `Wtf8Buf`

thaliaarchi opened the pull request:
This adds two more fast paths for UTF-8 validation in `Wtf8Buf`, making use of the `is_known_utf8` flag added in https://github.com/rust-lang/rust/pull/96869 (Optimize `Wtf8Buf::into_string` for the case where it contains UTF-8).

r? @ChrisDenton

thaliaarchi committed:
Add fast path for displaying pre-validated Wtf8Buf

thaliaarchi committed:
Skip scanning for surrogates when not known valid

ChrisDenton commented on library/std/src/sys_common/wtf8.rs:
```diff
@@ -166,6 +166,17 @@ impl fmt::Debug for Wtf8Buf {
+/// character, U+FFFD.
+impl fmt::Display for Wtf8Buf {
+    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if let Some(s) = self.as_known_utf8() {
+            return fmt::Display::fmt(s, formatter);
+        }
+        fmt::Display::fmt(&**self, formatter)
```
This could just be an `if`/`else` rather than early return

ChrisDenton reviewed (commented):
I only have one very small nit but otherwise it looks good to me.

thaliaarchi commented:
Now that https://github.com/rust-lang/rust/pull/137155 is finally through the merge queue, I've addressed your feedback and rebased onto it to resolve conflicts.

ChrisDenton commented:
Ok, looks good!

@bors r+

bors commented:
:pushpin: Commit eb14652770451022d424a1c301a4416514464932 has been approved by `ChrisDenton`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=ChrisDenton eb14652770451022d424a1c301a4416514464932 -->
<!-- homu: {"type":"Approved","sha":"eb14652770451022d424a1c301a4416514464932","approver":"ChrisDenton","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

ChrisDenton commented:
@bors rollup because it's small
